### ISSUE TITLE
Open tabs don't sync across devices. 

### ISSUE THREAD

**Issue by jtheletter** [brave/brave-browser#4412]
My bookmarks sync from desktop browser to iOS browser. But I see no way to access open tabs from one device to the other. I'm not even sure where to look. Omnibar autocomplete? History? 

macOS: Version 0.63.55 Chromium: 74.0.3729.131 (Official Build) (64-bit)

iOS: Version 1.9.1 (19.04.02.12)

**Comment by simonsan** [comment]
Besides syncing of open tabs in general I would also love to see some kind of folder structure where you can move open tabs between device folders. 

The folder structure (derived from the “new page overview”) on any open client could be:

Open tabs on mobile devices
- Device 1
  - Open Tab 1
  - Open Tab 2
  - ...
- Device 2
  - Open Tab 1
  - Open Tab 2
  - ...

Open tabs on desktop devices
- Device 1
  - Open Tab 1
  - Open Tab 2
  - ...
- Device 2
  - Open Tab 1
  - Open Tab 2
  - ...

With this you could easily send tabs to the other devices "open tabs"-folder. All clients having the same tabs open would be nice from the workflow point of view. Each client having it's own synced "open tabs" folder on all of the clients would make it easy to move/copy open tabs between devices and you could easily "switch" to another working environment.

Could be also making sense to not really divide it by devices but by own created "working environments" like this:

Working Environment 1 (e.g. Project X)
  - Open Tab 1
  - Open Tab 2
  - ...

Working Environment 1 (e.g. Android Phone at Work)
  - Open Tab 1
  - Open Tab 2
  - ...

Working Environment 2 (e.g. Desktop Home)
  - Open Tab 1
  - Open Tab 2
  - ...

The second idea would also make it possible to include the first idea manually but with extending possibilities to your own workflow. Another thing would be how to design it on the desktop. I could imagine that those "working environments" could start as a normal tab. Clicking on it gives an overview of the tabs inside this environment. Double clicking opens the working environment. Changing between the environments could be done by special static/solid tab left next to all the tabs in the tab bar where you could open different environments from a drop down menu either in a new window or in the same window, saving and closing the already opened environment. Moving tabs between environments could be done by right-clicking on a tab and a "send to environment --> list of possible environments" to send more then one tab you could shift-select more tabs and right-click one of them to send all via drop-down menu.

**Comment by s-r-aman** [comment]
+1 Would love to see this feature.

**Comment by rebron** [comment]
cc: @anthonypkeane

**Comment by simonsan** [comment]
> The folder structure seems great, would love to see it implemented. But it might introduce duplicate bookmarks.
> 
> Since the users may open bookmarked websites, by treating opened tabs as a special bookmark folder, you are (re)adding the tabs to bookmarks.

Basically using the word "bookmarks" was just an idea how to handle it internally to use the already in use sync engine for bookmarks.. The user itself should not see them as "bookmarks", but as "opened tabs".

**Comment by jtheletter** [comment]
Cross-device tab syncing is already a feature on Chrome. To be sure, I didn't mean to suggest a re-invented wheel.

**Comment by simonsan** [comment]
Restricting ad blockers to enterprise environments will be also a feature in the future of Chrome, but does that also mean you want to see that in Brave?

**Comment by jtheletter** [comment]
> Restricting ad blockers to enterprise environments…

I don't see the relevance of this question.

**Comment by simonsan** [comment]
It should point into the direction, that things Google does are hard to get rid of if they go that deeply in your infrastructure, because you also you get more dependent on them. So I personally could understand if you try to re-invent the wheel, you never want to get dependent on that privacy invader. Less Google, less amazon, less multinational concern structures make a better world, go reinvent the wheel, don't rely on Google.

**Comment by bsclifton** [comment]
@simonsan I'm going to collapse the comments since your question is off-topic... but to answer your specific question, please check out https://www.cnet.com/news/google-holds-firm-on-chrome-changes-that-may-break-ad-blockers/
> Brave, also based on Chromium, has ad and tracker blocking built in and enabled by default and thus will block that content regardless of Google's decisions. But Brave Chief Executive Brendan Eich said Thursday the company will maintain Chrome's older extension interface so uBlock Origin and another blocker from Hill, uMatrix, will continue to work.
>
> "We will support uBO and uMatrix," Eich said.

**Comment by kangarooo** [comment]
OMG. How then to export it all so i can continue working? How to save and export all tabs?

**Comment by Snuupy** [comment]
@kangarooo what I've been using personally is this: https://android.stackexchange.com/questions/56635/how-can-i-export-the-list-of-open-chrome-tabs

tl;dr Remote Devices feature in the Chrome debugger -> copy/paste -> your tabs are exported! :)

**Comment by simonsan** [comment]
> @kangarooo what I've been using personally is this: https://android.stackexchange.com/questions/56635/how-can-i-export-the-list-of-open-chrome-tabs
> 
> tl;dr Remote Devices feature in the Chrome debugger -> copy/paste -> your tabs are exported! :)

I added an answer there to manually export opened tabs from mobile android devices:
https://android.stackexchange.com/a/214518

While I had **650 open tabs** on my smartphone that needed to be manually "synchronized", I want to add a point to this issue:
Please **don't implement** something that automatically opens your tabs from another device. I think it makes sense to be able to chose what tab exactly you want to open. After doing all of this by hand, I'm even more convinced for a solution looking like [my proposal](https://github.com/brave/brave-browser/issues/4412#issuecomment-493698727) in this issue.

**Comment by itstheandre** [comment]
Hello! 

Is there any news on this? Firefox also has their send to device. Seeing that they are open source, couldnt this be something relatively easy to add?

**Comment by sabbir360** [comment]
> Hello!
> 
> Is there any news on this? Firefox also has their send to device. Seeing that they are open source, couldnt this be something relatively easy to add?

I think Chromium already have sync (History, Tabs, Bookmarks) which can be converted though this sync. stuff struggling from a long time. Have to be something quick to make it usable.

**Comment by ruifung** [comment]
Seriously, this is *still* not in yet? Like, this is the ONE feature stopping me from switching completely to brave.

**Comment by simonsan** [comment]
Wild guess: It might be due to #5198 that it's still not implemented and if, it might be in brave-core then.

**Comment by kangarooo** [comment]
Ok, at least then new Feature would solve- an option of button click to bookmark all opened tabs.
#5711 
https://github.com/brave/brave-browser/issues/5711

**Comment by simonsan** [comment]
Any news on this one? @bsclifton @rebron

Seems like this feature holds off some people to switch to or makes them switch away from Brave.
https://community.brave.com/t/recent-open-tab-sync-across-devices/49305/

**Comment by bsclifton** [comment]
No news on this yet, unfortunately - I agree, this would definitely be a nice feature! 😄

**Comment by simonsan** [comment]
I thought a long time about this, because I liked Brave, but one reason for this was the lack of this pretty basic feature. But also the prioritisation of this feature in comparison to other features (Crypto wallets? brave rewards? Who needs this stuff if not even the basic functionality is working?). So I'm back to my beloved **Firefox** with tab sync across all devices and to **Duck Duck Go privacy browser**. Goodbye Brave, we had some good and bad times together - now I'm gone. :-)

**Comment by thorinmesser** [comment]
It’s been over a year now since they released bookmark sync, which was claimed by the dev team to let tab sync come “much quicker” after it was done. I’d say those of us who need this feature are being forced to forget about brave and go back to firefox or chrome. It’s too bad, I really wanted to be part of this and use something I believe in so much. But they keep spending time and resources on weird niche features instead of finishing up the last few basics that all the competition has had for years now.

**Comment by sabifa** [comment]
Any update on this? Really missing this helpful feature 😢

**Comment by bsclifton** [comment]
cc: @anthonypkeane @jsecretan

**Comment by insilications** [comment]
Any news about this?

**Comment by vigneshr23** [comment]
I am waiting for this feature to ship. Syncing browser tabs and search history across chained devices is gonna make this browser great!

**Comment by stevewhitmore** [comment]
I think this may be a huge blocker for mass adoption. It seems to be the only thing that stops people from switching from Chrome when I tell them about Brave. Some can see past it but I think if you guys implement this feature you'll see a big increase in new users.

**Comment by skaldk** [comment]
> Could be also making sense to not really divide it by devices but by own created "working environments" like this:
> 
> Working Environment 1 (e.g. Project X)
> 
> * Open Tab 1
> * Open Tab 2
> * ...
> 
> Working Environment 1 (e.g. Android Phone at Work)
> 
> * Open Tab 1
> * Open Tab 2
> * ...
> 
> Working Environment 2 (e.g. Desktop Home)
> 
> * Open Tab 1
> * Open Tab 2
> * ...

Makes me feel like the Container plug-in on Firefox allowing you to have multiple tabs with open in the same windows but still related to their own session.

In my case : 
- admin access to my NAS
- my user access to my NAS
- Facebook, Twitter, sensitive data thief 

You may define urls to be opened in one container or another everytime and having a "not contained" session also for regular stuff. 

The good point is the fact that workspaces/containers are unconnected from each others (no access to the cookies from another workspace, or the list of opened tab).

**Comment by jsecretan** [comment]
Thanks everybody, we are actively working on a version 2 of sync right now, and synced tabs on devices is near the top of our list to include.

**Comment by sanjitagion** [comment]
Are there any extensions which serve this purpose in the meanwhile?

**Comment by jsecretan** [comment]
Confirmed that in the new Sync v2, you'll be able to see tabs from and send tabs to other devices.

**Comment by ckujau** [comment]
Why has this the `wontfix` label, I thougt this will be fixed with [Sync v2](https://github.com/brave/brave-browser/wiki/Brave-Sync-v2)? Wouldn't something like `Fixed` be a better tag here?

I assume [PR 5294](https://github.com/brave/brave-core/pull/5294) is the one that enabled `Sync v2`, and it's [enabled in dev/beta channel](https://github.com/brave/brave-core/commit/f3e63d417d7faf5887e93843b290c18f22b1087b "Enable sync v2 on dev/beta channel") - yay!

**Comment by jsecretan** [comment]
Yes, sorry for the confusion @ckujau

**Comment by Zvezdin** [comment]
Is there any way to view the currently open tabs in the current beta, or there isn't yet a UI for that? Nothing can be seen in the history tab or anywhere else in the sync menu and settings, however opening the context menu in top right and hovering over the "History" option displays the last few tabs open by all devices, though no way to expand that list.

**Comment by btlechowski** [comment]
Verification passed on

Brave | 1.12.102 Chromium: 84.0.4147.89 (Official Build) dev (64-bit)
-- | --
Revision | 19abfe7bcba9318a0b2a6bc6634a67fc834aa592-refs/branch-heads/4147@{#852}
OS | Ubuntu 18.04 LTS

Verified `Open tabs` are synced across devices. 
The `Open tabs` can be found in `Hamburger Menu -> History`
![image](https://user-images.githubusercontent.com/34715963/88589262-26cf2d80-d059-11ea-9430-2fe76928b0fe.png)
![image](https://user-images.githubusercontent.com/34715963/88589482-862d3d80-d059-11ea-949a-78b3c479b450.png)

`brave://history/` does not show synced tabs. Encountered https://github.com/brave/brave-browser/issues/10473
Favicons are not synced. Logged https://github.com/brave/brave-browser/issues/10949

**Comment by srirambv** [comment]
This is blocked on Android due to #11077. However, opened tabs from desktop and other devices are sync'd and listed via recent tabs on Android

OnePlus | Samsung Tab A
-- | --
![image](https://user-images.githubusercontent.com/17010094/89877689-5b45fc00-dbb8-11ea-94ca-78cf849179d7.png) | ![image](https://user-images.githubusercontent.com/17010094/89877747-71ec5300-dbb8-11ea-8710-0bb674a91097.png)

**Comment by cauerego** [comment]
oh, i see... the name "sync open tabs" is quite misleading!

it's just keeping a list of open tabs, not each tab own history and not an actual tab being synced across devices... so when activating it in one device, the other devices on the chain can see the tabs list under "recent tabs" (yet another misnomer).

the more you know... 🤣

**Comment by tbremer** [comment]
I don't see a complete list of my tabs, for instance…

Device A:
Window 1: 7 Tabs
Window 2: 4 Tabs

Device B:
Window 1: 2 Tabs

On Device B if I try to open a tab from Device A on Device B I only see a subset of those Tabs.

**Comment by fredless** [comment]
@tbremer see #11183

**Comment by tbremer** [comment]
Thanks @fredless!

**Comment by rijnhard** [comment]
How do I do this on iOS?

**Comment by anthonypkeane** [comment]
> How do I do this on iOS?

iOS doesn't have this ability yet. History is next on the Sync list for iOS and either passwords or Open Tabs will be after that.

**Comment by HeroHann** [comment]
So just to clarify: Sync between Mac and iOS / iPad OS does not work for tabs even after years of dev work?
Brave aims to be a a fully featured browser (now even with its own search engine) but does not have this simple feature??
Tried to get it to work for hours now...VERY disappointing! I guess I will have to stay with safari where this SIMPLE feature is working more or less.

Can someone explain WHY it is not working? I mean if there is a good explanation I might could find my peace with it...
SO why is it not working?

**Comment by dpschen** [comment]
I found a way to export the open tabs from Brave iOS:

1) Download and install [iMazing](https://imazing.com/) and create a backup of your device. Free version is fine for this.
2) Open `File System` in the iMazing sidebar and navigate to Brave/Library/ApplicationSupport/".
3) There you can find a file named `Brave.sqlite`. Copy this to your mac with the `Copy to Mac` button.
![Screenshot 2021-06-29 at 10 20 50](https://user-images.githubusercontent.com/6173598/123763693-579c9000-d8c4-11eb-9046-55c8a8baba1f.png)

4) You can open the Brave.sqlite e.g. with a program like [DB Browser for SQLite](https://sqlitebrowser.org/).
    The tabs are accessible via:

    - `Browse Data`
    - Select `ZTABMO` Table 
    - `Save the table as currently displayed` (small icon next to print button)
    - `Export to CSV`. You can open the CSV with Numbers, Excel, etc.
       ![Screenshot 2021-06-29 at 10 30 43](https://user-images.githubusercontent.com/6173598/123764779-62a3f000-d8c5-11eb-8ad0-109dd87debf3.png)

    The url is in the `ZURL` column, the title in `ZTITLE`.

---

At last my iPhone feels snappy again and the battery lasts much longer.
Having no sync and also no way to export or bookmark all tabs implemented and then **not even** telling your users (e.g. a small note in the sync setting on iOS) is simply irresponsible. After solving this for me I am switching finally back to Firefox.

**Comment by GregTheHun** [comment]
Yeah, this is a killer feature I would love to have. This is the one feature missing that makes me consider from time to time going back to Firefox. As long as Firefox is acting the way it is, I really don't want to, but it is something I would use all the time. I know the devs would too if they have to eat their own dog food and use this browser.

**Comment by Germs2004** [comment]
This apparently added support for every device type except iOS (which is the one the OP requested).  To vote for iOS to support syncing Open Tabs, go to [https://github.com/brave/brave-ios/issues/3229](https://github.com/brave/brave-ios/issues/3229).

**Comment by thorinmesser** [comment]
How can this be closed when it didn't implement the feature requested?

**Comment by bsclifton** [comment]
@thorinmesser it works for Android <=> Desktop; as shared above, iOS is tracked with https://github.com/brave/brave-ios/issues/3229

**Comment by jsecretan** [comment]
Yes, the team is working on this now for syncing between iOS and desktop.
